File: android/src/main/kotlin/io/intheloup/beacons/channel/Codec.kt
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0

package io.intheloup.beacons.channel

import com.squareup.moshi.Moshi
import io.intheloup.beacons.data.MonitoringState
import io.intheloup.beacons.data.Permission
import io.intheloup.beacons.data.Result
import io.intheloup.beacons.data.Settings

object Codec {

    private val moshi: Moshi = Moshi.Builder()
            .add(Permission.Adapter())
            .add(MonitoringState.Adapter())
            .add(Settings.Logs.Adapter())
            .build()

    fun encodeResult(result: Result): String =
            moshi.adapter(Result::class.java).toJson(result)

    fun decodePermission(arguments: Any?): Permission =
            Permission.Adapter().fromJson(arguments!! as String)

    fun decodeDataRequest(arguments: Any?): DataRequest =
            moshi.adapter(DataRequest::class.java).fromJson(arguments!! as String)!!

    fun decodeStatusRequest(arguments: Any?): StatusRequest =
            moshi.adapter(StatusRequest::class.java).fromJson(arguments!! as String)!!

    fun decodeSettings(arguments: Any?): Settings =
            moshi.adapter(Settings::class.java).fromJson(arguments!! as String)!!

}

File: ios/Classes/Channel/Codec.swift
//
//  Copyright (c) 2018 Loup Inc.
//  Licensed under Apache License v2.0
//

import Foundation

struct Codec {
  private static let jsonEncoder = JSONEncoder()
  private static let jsonDecoder = JSONDecoder()
  
  static func encode(result: Result) -> String {
    return String(data: try! jsonEncoder.encode(result), encoding: .utf8)!
  }
  
  static func encode(backgroundMonitoringEvent event: BackgroundMonitoringEvent) -> String {
    return String(data: try! jsonEncoder.encode(event), encoding: .utf8)!
  }
  
  static func decodePermission(from arguments: Any?) -> Permission {
    return Permission(rawValue: arguments! as! String)!
  }
  
  static func decodeStatusRequest(from arugments: Any?) -> StatusRequest {
    return try! jsonDecoder.decode(StatusRequest.self, from: (arugments as! String).data(using: .utf8)!)
  }
  
  static func decodeDataRequest(from arugments: Any?) -> DataRequest {
    return try! jsonDecoder.decode(DataRequest.self, from: (arugments as! String).data(using: .utf8)!)
  }
}
